src/components/tooltip/tooltip.js

```diff
@@ -10,6 +37,123 @@ import {h, Component} from 'preact';
  * @extends {Component}
  */
 class Tooltip extends Component {
+  state: Object;
+  _hoverTimeout: ?number;
+  textElement: HTMLSpanElement;
+  lastAlternativeTypeIndex: number = -1;
+
+  /**
+   * default component props
+   * @type {Object}
+   * @memberof Tooltip
+   */
+  static defaultProps = {
+    type: ToolTipType.Top,
+    maxWidth: '240px'
+  };
+
+  /**
+   * @constructor
+   * @param {*} props props
+   * @memberof Tooltip
+   */
+  constructor(props: Object) {
+    super(props);
+  }
+
+  /**
+   * on mouse over handler.
+   * @memberof Tooltip
+   * @returns {void}
+   */
+  onMouseOver(): void {
+    if (this._hoverTimeout) {
+      clearTimeout(this._hoverTimeout);
+      this._hoverTimeout = null;
+    }
+    this._hoverTimeout = setTimeout(() => {
+      this.setState({showTooltip: true});
+    }, TOOLTIP_SHOW_TIMEOUT);
+  }
+
+  /**
+   * on mouse leave handler.
+   * @memberof Tooltip
+   * @returns {void}
+   */
+  onMouseLeave(): void {
+    this.setState({showTooltip: false});
+    clearTimeout(this._hoverTimeout);
+    this._hoverTimeout = null;
+  }
+
+  /**
+   * brings another tooltip type which hasn't been marked as invalid
+   * @memberof Tooltip
+   * @returns {string} tooltip type
+   */
+  getAlternateType(): ?string {
+    return ((Object.values(ToolTipType).find((item, index) => {
+      if (index > this.lastAlternativeTypeIndex && item != this.props.type) {
+        this.lastAlternativeTypeIndex = index;
+        return true;
+      } else {
+        return false;
+      }
+    }): any): string);
+  }
+
+  /**
+   * checks if the current tooltip type is within the player boundaries
+   * @memberof Tooltip
+   * @returns {string} is in boundaries
+   */
+  isToolTipInBoundaries(): boolean {
+    const tooltipBoundingRect = this.textElement.getBoundingClientRect();
+    const playerContainerRect = this.props.playerClientRect;
+
+    return (
+      tooltipBoundingRect.top > playerContainerRect.top + PLAYER_MARGIN &&
+      tooltipBoundingRect.bottom < playerContainerRect.bottom - PLAYER_MARGIN &&
+      tooltipBoundingRect.right < playerContainerRect.right - PLAYER_MARGIN &&
+      tooltipBoundingRect.left > playerContainerRect.left + PLAYER_MARGIN
+    );
+  }
+
+  /**
+   * sets the requested type prop of the tooltip as a state cause it can change if is not valid
+   * @memberof Tooltip
+   * @returns {void}
+   */
+  componentWillMount(): void {
+    this.setState({valid: false, type: this.props.type});
+  }
+
+  /**
+   * checks if after the render the tooltip is within boundaries of the player
+   * if not it will try to set a new type which will be checked after the next render
+   * @param {Object} prevProps - previous component props
+   * @memberof Tooltip
+   * @returns {void}
+   */
+  componentDidUpdate(prevProps: Object): void {
+    if (this.props.playerClientRect !== prevProps.playerClientRect) {
+      this.lastAlternativeTypeIndex = -1;
+      this.setState({type: this.props.type});
```

Comment:
```suggestion
      this.setState({valid: false, type: this.props.type});
```
as you said yesterday. no?

Reply:
yes, you actually listen to what i say :))) thanks!